packages/ics721/src/types.rs

```diff
@@ -0,0 +1,68 @@
+use cosmwasm_schema::{cw_serde, schemars::JsonSchema};
+use cosmwasm_std::Binary;
+use serde::{Deserialize, Serialize};
+
+use crate::{ibc::NonFungibleTokenPacketData, token_types::ClassId};
+
+#[derive(Serialize, Deserialize, JsonSchema, Clone, Debug, PartialEq)]
+#[allow(clippy::derive_partial_eq_without_eq)]
+#[schemars(crate = "cosmwasm_schema::schemars")]
+#[serde(crate = "cosmwasm_schema::serde")]
+pub struct Ics721Memo {
+    pub callbacks: Option<Ics721Callbacks>,
+}
+
+/// The format we expect for the memo field on a send
+#[cw_serde]
+pub struct Ics721Callbacks {
+    /// Data to pass with a callback on source side (status update)
+    /// Note - If this field is empty, no callback will be sent
+    pub src_callback_msg: Option<Binary>,
+    /// The address that will receive the callback message
+    pub src_msg_receiver: Option<String>,
+    /// Data to pass with a callback on the destination side (ReceiveNftIcs721)
+    /// Note - If this field is empty, no callback will be sent
+    pub dest_callback_msg: Option<Binary>,
+    /// The address that will receive the callback message
+    pub dest_msg_receiver: Option<String>,
+}
+
+/// A message is that is being called on receiving the NFT after transfer was completed.
+/// Receiving this message means that the NFT was successfully transferred.
+/// You must verify this message was called by an approved ICS721 contract, either by code_id or address.
+#[cw_serde]
+pub struct Ics721ReceiveMsg {
+    pub original_packet: NonFungibleTokenPacketData,
+    pub local_class_id: ClassId,
+    pub msg: Binary,
+}
+
+/// A message to update your contract of the status of a transfer
+/// status = Ics721Status::Success - the transfer was successful and NFT is on the other chain
+/// status = Ics721Status::Failed - Transfer failed and contract still owns the NFT
+#[cw_serde]
+pub struct Ics721CallbackMsg {
+    pub status: Ics721Status,
+    pub original_packet: NonFungibleTokenPacketData,
+    pub msg: Binary,
+}
+
+/// The status of a transfer on callback
+#[cw_serde]
+pub enum Ics721Status {
+    Success,
+    Failed,
+}
+
+/// This is a wrapper for ics721 callbacks
+/// so contracts will be able to recieve both status update and on receive hook.
+#[cw_serde]
+pub enum ReceiverExecuteMsg {
+    /// Being called on receiving the NFT after transfer was completed. (destination side)
+    /// `on_recieve` hook
+    /// Note - Failing this message will fail the transfer.
+    ReceiveNftIcs721(Ics721ReceiveMsg),
```

Comment:
Would rename this to Ics721ReceiveNft.

Would also rename:
 Ics721Callback ->  Ics721CallbackNft
Ics721ReceiveMsg -> Ics721ReceiveNftMsg
Ics721CallbackMsg -> Ics721CallbackNftMsg

Would make naming more consistent.

Out of consistenty

Reply:
What you think about this: 
```pub enum ReceiverExecuteMsg {
    Ics721ReceiveCallback(Ics721ReceiveCallbackMsg),
    Ics721AckCallback(Ics721AckCallbackMsg),
}```

I want to make it clear we have 2 type of callbacks (hooks).
1. Receive callback: is being called on the target (receiver) chain, after NFT was transferred successfully. 
2. Ack callback: is being called on the source (sender) chain, after transfer, letting the contract know of the status of the transfer (succeed or failed).

Reply:
Perfect!